crates/compression/src/tables.rs

```diff
@@ -0,0 +1,210 @@
+};
+use std::collections::HashMap;
+
+use crate::{
+    compress::{
+        CompressCtx,
+        CompressDb,
+        PrepareCtx,
+    },
+    decompress::{
+        DecompressCtx,
+        DecompressDb,
+        DecompressError,
+    },
+    ports::{
+        EvictorDb,
+        TemporalRegistry,
+    },
+};
+
+macro_rules! tables {
+    ($($name:ident: $type:ty),*$(,)?) => {
+        #[doc = "RegistryKey namespaces"]
+        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, serde::Serialize, serde::Deserialize)]
+        #[allow(non_camel_case_types)] // Match names in structs exactly
+        pub enum RegistryKeyspace {
+            $(
+                $name,
+            )*
+        }
+
+        #[doc = "RegistryKey namespace with an associated typed value"]
+        #[derive(Debug, Clone, PartialEq, Eq, Hash, serde::Serialize, serde::Deserialize)]
+        #[allow(non_camel_case_types)] // Match names in structs exactly
+        pub enum RegistryKeyspaceValue {
+            $(
+                $name($type),
+            )*
+        }
+
+        #[doc = "A value for each keyspace"]
+        #[derive(Debug, Clone, Default)]
+        pub struct PerRegistryKeyspace<T> {
+            $(pub $name: T,)*
+        }
+        impl<T> core::ops::Index<RegistryKeyspace> for PerRegistryKeyspace<T> {
+            type Output = T;
+
+            fn index(&self, index: RegistryKeyspace) -> &Self::Output {
+                match index {
+                    $(
+                        RegistryKeyspace::$name => &self.$name,
+                    )*
+                }
+            }
+        }
+        impl<T> core::ops::IndexMut<RegistryKeyspace> for PerRegistryKeyspace<T> {
+            fn index_mut(&mut self, index: RegistryKeyspace) -> &mut Self::Output {
+                match index {
+                    $(
+                        RegistryKeyspace::$name => &mut self.$name,
+                    )*
+                }
+            }
+        }
+
+        #[doc = "Key-value mapping for each keyspace"]
+        #[derive(Debug, Clone, Default)]
+        pub struct PerRegistryKeyspaceMap {
+            $(pub $name: HashMap<RegistryKey, $type>,)*
+        }
+
+        impl PerRegistryKeyspaceMap {
+            #[cfg(test)]
+            pub fn insert(&mut self, key: RegistryKey, value: RegistryKeyspaceValue) {
+                match value {
+                    $(
+                        RegistryKeyspaceValue::$name(value) => {
+                            self.$name.insert(key, value);
+                        }
+                    )*
+                }
+            }
+        }
+
+        #[doc = "The set of registrations for each table, as used in the compressed block header"]
+        #[derive(Debug, Clone, PartialEq, Default, serde::Serialize, serde::Deserialize)]
+        pub struct RegistrationsPerTable {
+            $(pub $name: Vec<(RegistryKey, $type)>,)*
+        }
+
+        impl TryFrom<PerRegistryKeyspaceMap> for RegistrationsPerTable {
```

Comment:
Why do we implement `TryFrom` instead of `From`?

Reply:
It used to be fallible when we used serde. Now no reason, so I'll change it.

Reply:
Fixed in 8d5fdf20c5d55347dee3724f823d3b1479de5d07